File: experiments/adrianco/experiment-47-gptoss20b/bookshop/runs/agent=hermes-local_language=go_model=mlxlocal/mlx-community--gpt-oss-20b-MXFP4-Q8_prompt=neutral_stack=gptoss/rep3-failed/router.go
package main

import (
    "database/sql"
    "encoding/json"
    "net/http"
    "strconv"
    "strings"

    "github.com/go-chi/chi/v5"
)

// Book represents a book record.
type Book struct {
    ID     int64  `json:"id"`
    Title  string `json:"title"`
    Author string `json:"author"`
    Year   int    `json:"year,omitempty"`
    ISBN   string `json:"isbn,omitempty"`
}

func NewRouter() http.Handler {
    r := chi.NewRouter()

    r.Get("/health", healthHandler)

    r.Route("/books", func(r chi.Router) {
        r.Get("/", listBooksHandler)
        r.Post("/", createBookHandler)
        r.Get("/{id}", getBookHandler)
        r.Put("/{id}", updateBookHandler)
        r.Delete("/{id}", deleteBookHandler)
    })
    return r
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

func createBookHandler(w http.ResponseWriter, r *http.Request) {
    var b Book
    if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
        http.Error(w, "invalid json", http.StatusBadRequest)
        return
    }
    if strings.TrimSpace(b.Title) == "" || strings.TrimSpace(b.Author) == "" {
        http.Error(w, "title and author required", http.StatusBadRequest)
        return
    }
    db := getDB()
    res, err := db.Exec("INSERT INTO books (title, author, year, isbn) VALUES (?, ?, ?, ?)", b.Title, b.Author, b.Year, b.ISBN)
    if err != nil {
        http.Error(w, "db error", http.StatusInternalServerError)
        return
    }
    id, _ := res.LastInsertId()
    b.ID = id
    w.Header().Set("Content-Type", "application/json")
    w.WriteHeader(http.StatusCreated)
    json.NewEncoder(w).Encode(b)
}

func listBooksHandler(w http.ResponseWriter, r *http.Request) {
    author := r.URL.Query().Get("author")
    db := getDB()
    var rows *sql.Rows
    var err error
    if author != "" {
        rows, err = db.Query("SELECT id, title, author, year, isbn FROM books WHERE author = ?", author)
    } else {
        rows, err = db.Query("SELECT id, title, author, year, isbn FROM books")
    }
    if err != nil {
        http.Error(w, "db error", http.StatusInternalServerError)
        return
    }
    defer rows.Close()
    books := []Book{}
    for rows.Next() {
        var b Book
        if err := rows.Scan(&b.ID, &b.Title, &b.Author, &b.Year, &b.ISBN); err != nil {
            continue
        }
        books = append(books, b)
    }
    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(books)
}

func getBookHandler(w http.ResponseWriter, r *http.Request) {
    idStr := chi.URLParam(r, "id")
    id, err := strconv.ParseInt(idStr, 10, 64)
    if err != nil {
        http.Error(w, "invalid id", http.StatusBadRequest)
        return
    }
    db := getDB()
    row := db.QueryRow("SELECT id, title, author, year, isbn FROM books WHERE id = ?", id)
    var b Book
    if err := row.Scan(&b.ID, &b.Title, &b.Author, &b.Year, &b.ISBN); err != nil {
        http.Error(w, "not found", http.StatusNotFound)
        return
    }
    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(b)
}

func updateBookHandler(w http.ResponseWriter, r *http.Request) {
    idStr := chi.URLParam(r, "id")
    id, err := strconv.ParseInt(idStr, 10, 64)
    if err != nil {
        http.Error(w, "invalid id", http.StatusBadRequest)
        return
    }
    var b Book
    if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
        http.Error(w, "invalid json", http.StatusBadRequest)
        return
    }
    if strings.TrimSpace(b.Title) == "" || strings.TrimSpace(b.Author) == "" {
        http.Error(w, "title and author required", http.StatusBadRequest)
        return
    }
    db := getDB()
    res, err := db.Exec("UPDATE books SET title = ?, author = ?, year = ?, isbn = ? WHERE id = ?", b.Title, b.Author, b.Year, b.ISBN, id)
    if err != nil {
        http.Error(w, "db error", http.StatusInternalServerError)
        return
    }
    if rows, _ := res.RowsAffected(); rows == 0 {
        http.Error(w, "not found", http.StatusNotFound)
        return
    }
    b.ID = id
    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(b)
}

func deleteBookHandler(w http.ResponseWriter, r *http.Request) {
    idStr := chi.URLParam(r, "id")
    id, err := strconv.ParseInt(idStr, 10, 64)
    if err != nil {
        http.Error(w, "invalid id", http.StatusBadRequest)
        return
    }
    db := getDB()
    res, err := db.Exec("DELETE FROM books WHERE id = ?", id)
    if err != nil {
        http.Error(w, "db error", http.StatusInternalServerError)
        return
    }
    if rows, _ := res.RowsAffected(); rows == 0 {
        http.Error(w, "not found", http.StatusNotFound)
        return
    }
    w.WriteHeader(http.StatusNoContent)
}

File: experiments/adrianco/experiment-56-terra-alllang/bookshop/runs/agent=codex_effort=default_language=elixir_model=gpt-5.6-terra_prompt=neutral/rep1/lib/book_api/router.ex
defmodule BookApi.Router do
  @moduledoc false

  def route("GET", "/health", _query, _body), do: {200, %{status: "ok"}}
  def route("GET", "/books", query, _body), do: {200, BookApi.Store.list(Map.get(query, "author"))}
  def route("GET", "/books/" <> id, _query, _body), do: fetch(id)
  def route("POST", "/books", _query, body), do: with_book(body, &{201, BookApi.Store.create(&1)})
  def route("PUT", "/books/" <> id, _query, body), do: with_book(body, fn book -> update(id, book) end)
  def route("DELETE", "/books/" <> id, _query, _body), do: delete(id)
  def route(_, _, _, _), do: {404, %{error: "not found"}}

  defp fetch(id) do
    case valid_id(id) do
      {:ok, book_id} -> case BookApi.Store.get(book_id) do nil -> not_found(); book -> {200, book} end
      :error -> not_found()
    end
  end
  defp update(id, book) do
    case valid_id(id) do
      {:ok, book_id} -> case BookApi.Store.update(book_id, book) do nil -> not_found(); updated -> {200, updated} end
      :error -> not_found()
    end
  end
  defp delete(id) do
    case valid_id(id) do
      {:ok, book_id} -> case BookApi.Store.delete(book_id) do nil -> not_found(); _ -> {204, nil} end
      :error -> not_found()
    end
  end
  defp not_found, do: {404, %{error: "book not found"}}

  defp with_book(body, action) do
    with {:ok, params} <- BookApi.JSON.decode(body),
         {:ok, book} <- validate(params) do
      action.(book)
    else
      {:error, reason} -> {422, %{error: reason}}
    end
  end
  defp validate(params) when is_map(params) do
    title = Map.get(params, "title")
    author = Map.get(params, "author")
    year = Map.get(params, "year")
    isbn = Map.get(params, "isbn")
    cond do
      not (is_binary(title) and String.trim(title) != "") -> {:error, "title is required"}
      not (is_binary(author) and String.trim(author) != "") -> {:error, "author is required"}
      not (is_nil(year) or is_integer(year)) -> {:error, "year must be an integer"}
      not (is_nil(isbn) or is_binary(isbn) or is_integer(isbn)) -> {:error, "isbn must be a string"}
      true -> {:ok, %{title: String.trim(title), author: String.trim(author), year: year, isbn: if(is_nil(isbn), do: nil, else: to_string(isbn))}}
    end
  end
  defp validate(_), do: {:error, "request body must be a JSON object"}
  defp valid_id(id) do
    case Integer.parse(id) do {value, ""} when value > 0 -> {:ok, value}; _ -> :error end
  end
end
